src/librustc_lint/builtin.rs

```diff
@@ -173,6 +173,12 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for NonShorthandFieldPatterns {
                 }
                 if let PatKind::Binding(_, _, ident, None) = fieldpat.node.pat.node {
                     if ident.node == fieldpat.node.name {
+                        if let Some(_) = fieldpat.span.ctxt().outer().expn_info() {
```

Comment:
`Value { $a: value }` is also possible, so checking that *any* of the spans is from a macro would be more correct.

(The proper way to fix this issue is to use hygienic comparison, but we don't have necessary information available in HIR right now.)

Reply:
> we don't have necessary information available in HIR right now

Is this planned or being worked on? (If there's a plan, we can add a `# FIXME: do hygenic comparison when this-and-such is complete` comment.)

Reply:
There's a plan, I'm working on some identifier/span changes that would include the fix for this issue and I was aware about it before #49588 was submitted, but there's no issue that can be referred to in FIXME.

Reply:
> `Value { $a: value }` is also possible

Um, I'm having trouble coming up with an example. Where could the variable `value` come from? It doesn't seem like I can define it earlier in the macro like

```rust
macro_rules! pat_again {
    ($a:pat) => {
        let value = 2;
        Value { $a: value }
    };
}
```

because the entire macro needs to expand into a pattern. What am I missing? (Regrettably, I have not heretofore written very many macros.)

Reply:
Wait, nevermind; I think I'm anchoring too hard on the given example

Reply:
(not well-thought-out comment deleted)